File: cpp/test/phonenumbers/phonecontextparser_test.cc
// Copyright (C) 2025 The Libphonenumber Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "phonenumbers/phonecontextparser.h"

#include <gtest/gtest.h>

#include "phonenumbers/phonenumbernormalizer.h"
#include "phonenumbers/regexpsandmappings.h"

namespace i18n {
namespace phonenumbers {
using testing::Eq;

class PhoneContextParserTest : public testing::Test {
 public:
  // This type is neither copyable nor movable.
  PhoneContextParserTest(const PhoneContextParserTest&) = delete;
  PhoneContextParserTest& operator=(const PhoneContextParserTest&) = delete;

 protected:
  PhoneContextParserTest()
      : country_calling_codes_(
            std::make_unique<std::vector<int>>(std::vector<int>{64})),
        reg_exps_(new PhoneNumberRegExpsAndMappings()),
        normalizer_(new PhoneNumberNormalizer(reg_exps_)),
        context_parser_(new PhoneContextParser(
            std::move(country_calling_codes_), reg_exps_, normalizer_)) {}

  std::unique_ptr<std::vector<int>> country_calling_codes_;
  std::shared_ptr<PhoneNumberRegExpsAndMappings> reg_exps_;
  std::shared_ptr<PhoneNumberNormalizer> normalizer_;
  std::unique_ptr<PhoneContextParser> context_parser_;

  absl::StatusOr<std::optional<PhoneContextParser::PhoneContext>> Parse(
      absl::string_view phone_number) {
    return context_parser_->Parse(phone_number);
  }
};

TEST_F(PhoneContextParserTest, ParsePhoneContext) {
  auto parse_result = Parse("tel:03-331-6005;phone-context=+64");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_TRUE(parse_result->has_value());
  EXPECT_EQ("+64", parse_result.value()->raw_context);
  EXPECT_EQ(64, parse_result.value()->country_code);

  auto parse_result = Parse("tel:03-331-6005;phone-context=example.com");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_TRUE(parse_result->has_value());
  EXPECT_EQ("example.com", parse_result.value()->raw_context);
  EXPECT_EQ(std::nullopt, parse_result.value()->country_code);

  auto parse_result = Parse("03-331-6005;phone-context=+64;");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_TRUE(parse_result->has_value());
  EXPECT_EQ("+64", parse_result.value()->raw_context);
  EXPECT_EQ(64, parse_result.value()->country_code);

  auto parse_result = Parse("+64-3-331-6005;phone-context=+64;");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_TRUE(parse_result->has_value());
  EXPECT_EQ("+64", parse_result.value()->raw_context);
  EXPECT_EQ(64, parse_result.value()->country_code);

  auto parse_result =
      Parse("tel:03-331-6005;foo=bar;phone-context=+64;baz=qux");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_TRUE(parse_result->has_value());
  EXPECT_EQ("+64", parse_result.value()->raw_context);
  EXPECT_EQ(64, parse_result.value()->country_code);

  auto parse_result = Parse("tel:03-331-6005");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_EQ(std::nullopt, parse_result);

  auto parse_result = Parse("tel:03-331-6005;phone-context=+0");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_TRUE(parse_result->has_value());
  EXPECT_EQ("+0", parse_result.value()->raw_context);
  EXPECT_EQ(std::nullopt, parse_result.value()->country_code);

  auto parse_result = Parse("tel:03-331-6005;phone-context=+1234");
  ASSERT_TRUE(parse_result.ok());
  ASSERT_TRUE(parse_result->has_value());
  EXPECT_EQ("+1234", parse_result.value()->raw_context);
  EXPECT_EQ(std::nullopt, parse_result.value()->country_code);
}

TEST_F(PhoneContextParserTest, ParsePhoneContextInvalid) {
  auto parse_result = Parse("tel:03-331-6005;phone-context=");
  EXPECT_EQ(absl::StatusCode::kInvalidArgument, parse_result.status().code());

  auto parse_result = Parse("tel:03-331-6005;phone-context=;");
  EXPECT_EQ(absl::StatusCode::kInvalidArgument, parse_result.status().code());

  auto parse_result = Parse("tel:03-331-6005;phone-context=0");
  EXPECT_EQ(absl::StatusCode::kInvalidArgument, parse_result.status().code());
}

}  // namespace phonenumbers
}  // namespace i18n

File: java/libphonenumber/test/com/google/i18n/phonenumbers/PhoneContextParserTest.java
/*
 * Copyright (C) 2025 The Libphonenumber Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.google.i18n.phonenumbers;

import static org.junit.Assert.assertThrows;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNull;
import java.util.Collections;
import java.util.HashSet;
import java.util.Set;
import org.junit.function.ThrowingRunnable;
import com.google.i18n.phonenumbers.PhoneContextParser.PhoneContext;
import junit.framework.TestCase;

/**
 * Unit tests for PhoneContextParser.java
 */
public class PhoneContextParserTest extends TestCase {

  private static final Set<Integer> countryCallingCodeSet;
  static {
    Set<Integer> tempSet = new HashSet<>();
    tempSet.add(64);
    countryCallingCodeSet = Collections.unmodifiableSet(tempSet);
  }

  /**
   * An instance of PhoneContextParser.
   */
  protected final PhoneContextParser phoneContextParser =
      new PhoneContextParser(countryCallingCodeSet);

  public void testParseShouldWorkAsExpected() throws NumberParseException {
    PhoneContext actual;

    actual = phoneContextParser.parse("tel:03-331-6005;phone-context=+64");
    assertEquals("+64", actual.getRawContext());
    assertEquals(new Integer(64), actual.getCountryCode());
    
    actual = phoneContextParser.parse("tel:03-331-6005;phone-context=example.com");
    assertEquals("example.com", actual.getRawContext());
    assertNull(actual.getCountryCode());
    
    actual = phoneContextParser.parse("03-331-6005;phone-context=+64;");
    assertEquals("+64", actual.getRawContext());
    assertEquals(new Integer(64), actual.getCountryCode());
    
    actual = phoneContextParser.parse("+64-3-331-6005;phone-context=+64;");
    assertEquals("+64", actual.getRawContext());
    assertEquals(new Integer(64), actual.getCountryCode());
    
    actual = phoneContextParser.parse("tel:03-331-6005;foo=bar;phone-context=+64;baz=qux");
    assertEquals("+64", actual.getRawContext());
    assertEquals(new Integer(64), actual.getCountryCode());
    
    actual = phoneContextParser.parse("tel:03-331-6005");
    assertNull(actual);
    
    actual = phoneContextParser.parse("tel:03-331-6005;phone-context=+0");
    assertEquals("+0", actual.getRawContext());
    assertNull(actual.getCountryCode());
    
    actual = phoneContextParser.parse("tel:03-331-6005;phone-context=+1234");
    assertEquals("+1234", actual.getRawContext());
    assertNull(actual.getCountryCode());
  }

  public void testParseShouldFailForInvalidPhoneContext() throws NumberParseException {
    assertThrows(
        NumberParseException.class,
        new ThrowingRunnable() {
          @Override
          public void run() throws NumberParseException {
            phoneContextParser.parse("tel:03-331-6005;phone-context=");
          }
        });
    assertThrows(
        NumberParseException.class,
        new ThrowingRunnable() {
          @Override
          public void run() throws NumberParseException {
            phoneContextParser.parse("tel:03-331-6005;phone-context=;");
          }
        });
    assertThrows(
        NumberParseException.class,
        new ThrowingRunnable() {
          @Override
          public void run() throws NumberParseException {
            phoneContextParser.parse("tel:03-331-6005;phone-context=0");
          }
        });
  }

}
